geth/peers/topicpool_test.go

```diff
@@ -268,3 +268,39 @@ func (s *TopicPoolSuite) TestPeerDontAddTwice() {
 	// peer2 already added to p2p server no reason to add it again
 	s.Nil(s.topicPool.AddPeerFromTable(s.peer))
 }
+
+func (s *TopicPoolSuite) TestMaxCachedPeers() {
```

Comment:
why do you have to drop every peer?
i think that correct test wouldn't do that. just found a peer -> confirm added
and then found 2 more. after that 3 peers must be cached

it seems that in the case that i described those 2 wont be cached because of https://github.com/status-im/status-go/blob/develop/geth/peers/topicpool.go#L393

Reply:
@dshulyak but, if i understood well, entry point for adding new peers is on [p2p.PeerEventTypeAdd events](https://github.com/status-im/status-go/pull/989/files#diff-fb40a852d6fa4cad34711d85a9689097R241) which will be calling [t.ConfirmAdded](https://github.com/status-im/status-go/pull/989/files#diff-fb40a852d6fa4cad34711d85a9689097R258) for every new peer on every topic.

While `processFoundNode` will add the peer to `pendingPeers` map `ConfirmAdded` will get this map entry and add it to the cache with `t.cache.AddPeer`.

The original test was simulating this behaviour by running something like this:
```
func (s *TopicPoolSuite) TestMaxCachedPeers() {
	s.topicPool.limits = params.NewLimits(1, 1)
	s.topicPool.maxCachedPeers = 3
	peer1 := discv5.NewNode(discv5.NodeID{1}, s.peer.Self().IP, 32311, 32311)
	peer2 := discv5.NewNode(discv5.NodeID{2}, s.peer.Self().IP, 32311, 32311)
	peer3 := discv5.NewNode(discv5.NodeID{3}, s.peer.Self().IP, 32311, 32311)
	s.topicPool.processFoundNode(s.peer, peer1)
	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer1.ID))
	s.topicPool.processFoundNode(s.peer, peer2)
	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer2.ID))
	s.topicPool.processFoundNode(s.peer, peer3)
	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer3.ID))

	cached := s.topicPool.cache.GetPeersRange(s.topicPool.topic, 5)
	s.Equal(3, len(cached))

	s.True(s.topicPool.maxCachedReached())
}
```

Is there anything i'm missing?

Reply:
so at first node gets discovered and processFoundNode is called

if max limit is not reached this peer will be added to a server
if handshake in the server is successful -> confirm added is called and peer is persisted
if we added more than max then surplus is dropped but we don't remove those peers from cache

if max limit is reached peer is added only to a heap queue which is not persisted

test that you pasted in the comment looks correct

Reply:
sorry, it is not correct
because this line https://github.com/status-im/status-go/blob/develop/geth/peers/topicpool.go#L394 won't be executed and thus ConfirmAdded won't be called

Reply:
@dshulyak I probably misunderstood the pendingPeers vs. connectedPeers on first instance. 

So, peers found between maxPeers and maxCachedPeers should be part of `connectedPeers` instead of `pendingPeers` as they're confirmed server peers and not peers to be confirmed. My initial thought was to think on pendingPeers as cachedPeers which is not correct.

I think [this commit](https://github.com/status-im/status-go/pull/989/commits/580f07c0b1b2ef774ccc3d012a813cf170901312) will addresses the problem.

Reply:
sorry just noticed your comment

1. i am not sure why peers that were dropped now present in connectedPeers table. connectedPeers are only for peers that connected to mail server. caching them should not affect that table

2. this test shouldn't drop peers at all. because it breaks the logic of the test